mlir/lib/Quantum/Transforms/VerifyParentGateAnalysis.hpp

```diff
@@ -0,0 +1,139 @@
+// Copyright 2024 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This analysis checks if a gate operation and its parent gate operation
+// are correctly matched for the purposes of peephole optimizations like
+// merge rotations and cancel inverses.
+// Gates passing this analysis are considered valid candidates for merge
+// rotation and cancel inverses.
+
+// Specifically, we check the following conditions:
+//  1. Both gates must be of the same type, i.e. a quantum.custom can
+//     only be cancelled with a quantum.custom, not a quantum.unitary
+//  2. The results of the parent gate must map one-to-one, in order,
+//     to the operands of the second gate
+//     For example, the pair
+//        %0:2 = quantum.custom "CNOT"() %.., %..
+//        %1:2 = quantum.custom "CNOT"() %0#0, %0#1
+//     is considered to be a successful match, but the pair
+//        %0:2 = quantum.custom "CNOT"() %.., %..
+//        %1:2 = quantum.custom "CNOT"() %0#1, %0#0
+//     is not.
+//  3. If the gates are controlled, both gates' control wires and values
+//     must be the same. The control wires must be in the same order
+
+#pragma once
+
+#include "mlir/IR/BuiltinOps.h"
+#include "llvm/Support/Debug.h"
+
+#include "Catalyst/IR/CatalystDialect.h"
+#include "Quantum/IR/QuantumOps.h"
+
+using namespace llvm;
+using namespace mlir;
+using namespace catalyst;
+
+namespace catalyst {
+
+template <typename OpType> class VerifyParentGateAnalysis {
```

Comment:
Are all op type supported in the analysis? `GateOp`, `CustomOp`,  `UnitaryGateOp`, `SetStateOp`, `GlobalPhaseOp`, `MultiRZOp`, and more. I am not sure all the methods exist between those classes.

Reply:
I am looking at how to use the [existing quantum opinterface](https://github.com/PennyLaneAI/catalyst/blob/main/mlir/include/Quantum/IR/QuantumInterfaces.td) instead of adding template types manually

Reply:
Yes that makes sense!

Reply:
`GlobalPhase` is a bit different from the others since it doesn't actually have qubit operands/results. Let's skip that for now I think since it's not particularly important.

Reply:
Ok no worries :+1: